BHoM_UI/Templates/Selector.cs

```diff
@@ -71,26 +71,36 @@ public Selector(IEnumerable<T> possibleItems, string key)
 
         public void AddToMenu(ToolStripDropDown menu)
         {
-            if (m_Menu_WinForm == null)
-            {
-                m_Menu_WinForm = new SelectorMenu_WinForm<T>(m_ItemListStore[m_Key], m_ItemTreeStore[m_Key]);
-                m_Menu_WinForm.ItemSelected += M_Menu_ItemSelected;
-            }
+            if (m_SelectorMenu == null)
+                SetSelectorMenu(new SelectorMenu_WinForm<T>(m_ItemListStore[m_Key], m_ItemTreeStore[m_Key]));
 
-            m_Menu_WinForm.FillMenu(menu);
+            m_SelectorMenu.FillMenu(menu);
         }
 
         /*************************************/
 
         public void AddToMenu(System.Windows.Controls.ContextMenu menu)
         {
-            if (m_Menu_Wpf == null)
-            {
-                m_Menu_Wpf = new SelectorMenu_Wpf<T>(m_ItemListStore[m_Key], m_ItemTreeStore[m_Key]);
-                m_Menu_Wpf.ItemSelected += M_Menu_ItemSelected;
-            }
+            if (m_SelectorMenu == null)
+                SetSelectorMenu(new SelectorMenu_Wpf<T>(m_ItemListStore[m_Key], m_ItemTreeStore[m_Key]));
+
+            m_SelectorMenu.FillMenu(menu);
+        }
+
+        /*************************************/
 
-            m_Menu_Wpf.FillMenu(menu);
+        public void AddToMenu(object menu)
+        {
+            if (m_SelectorMenu != null)
+                m_SelectorMenu.FillMenu(menu);
+        }
+
+        /*************************************/
+
+        public void SetSelectorMenu<M>(SelectorMenu<T, M> selectorMenu)
```

Comment:
The problem with this is two-fold

1. outside of the context of `Selector<T>` or `Caller.SetPossibleItems<T>` we can't (statically) know what T is
2. outside of the context of `Selector<T>` we don't have access to the private members that make up the `SelectorMenu<T, M>` constructor parameters

The former could be fixed by a factory method that takes a `Selector<T>` as dynamic and returns something specialised on T, but then we still don't have the `List<SearchItem>, Tree<T>` to pass into the constructor.

My thinking is that we could have a factory method inside `Selector` (and declared in `ISelector`) that takes a `Type` at runtime (which has to be an incomplete type that is a subclass of `SelectorMenu<T, M>` partially specialised on `M` but not `T` that has a constructor that takes the arguments `List<SearchItem>, Tree<T>`) and invokes that constructor dynamically, then. But I really don't like that, it seems very inelegant.

The other idea I've had is to declare a `SetItems(List<SearchItem>, Tree<T>)` in `ISelectorMenu<T>` that will be called by `Selector<T>.SetSelectorMenu<M>()`, removing the need to construct a `SelectorMenu<T,M>` with those values and the library that creates the `SelectorMenu<T, M>` can implement a factory that takes a `Selector<T>`; creates its `SelectorMenu<T,M>` implementation, specialised on `T` and `M`; and calls `Selector<T>.SetSelectorMenu<M>()` on its parameter.

Any other ideas of how to handle this statically?